random/resource_string_migration.go

```diff
@@ -0,0 +1,48 @@
+package random
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func resourceRandomStringMigrateState(
+	v int, is *terraform.InstanceState, meta interface{}) (*terraform.InstanceState, error) {
+	switch v {
+	case 0:
+		log.Println("[INFO] Found random string state v0; migrating to v1")
+		return migrateStringStateV0toV1(is)
+	default:
+		return is, fmt.Errorf("Unexpected schema version: %d", v)
+	}
+}
+
+func redactAttributes(is *terraform.InstanceState) map[string]string {
+	redactedAttributes := make(map[string]string)
+	for k, v := range is.Attributes {
+		redactedAttributes[k] = v
+		if k == "id" || k == "result" {
+			redactedAttributes[k] = "<sensitive>"
+		}
+	}
+	return redactedAttributes
+}
+
+func migrateStringStateV0toV1(is *terraform.InstanceState) (*terraform.InstanceState, error) {
+	if is.Empty() {
+		log.Println("[DEBUG] Empty InstanceState; nothing to migrate.")
+		return is, nil
+	}
+
+	log.Printf("[DEBUG] Random String Attributes before Migration: %#v", redactAttributes(is))
+
+	is.Attributes["min_numeric"] = "0"
+	is.Attributes["min_upper"] = "0"
+	is.Attributes["min_lower"] = "0"
+	is.Attributes["min_special"] = "0"
```

Comment:
Because we've now shipped a version where these _did_ exist in schema version 0, I think for safety we need to handle these conditionally so that we don't inadvertently cause further problems for people who have already applied changes with the v1.3.0 release.

I think it should be sufficient to just apply this change conditionally only if the attribute is not already set:

```hcl
	if v := is.Attributes["min_special"]; v == "" {
		is.Attributes["min_special"] = "0"
	}
```

This would ensure that we don't overwrite any nonzero values that might have been inserted by users already trying to use the new feature with `v1.3.0`, which would then cause another forced replacement for those users.